Parser/tokenizer.c

```diff
@@ -370,6 +370,72 @@ update_fstring_buffers(struct tok_state *tok, char value, int regular, int multi
     }
 }
 
+static int
+update_fstring_expr(struct tok_state *tok, char cur)
+{
+    assert(tok->cur != NULL);
+
+    Py_ssize_t size = strlen(tok->cur);
+    tokenizer_mode *tok_mode = &(tok->tok_mode_stack[tok->tok_mode_stack_index]);
+
+    switch (cur) {
+        case '{':
+            if (tok_mode->last_expr_buffer != NULL) {
+                PyMem_Free(tok_mode->last_expr_buffer);
+            }
+            tok_mode->last_expr_buffer = PyMem_Malloc(size);
+            if (tok_mode->last_expr_buffer == NULL) {
+                tok->done = E_NOMEM;
+                return 0;
+            }
+            tok_mode->last_expr_size = size;
+            tok_mode->last_expr_end = -1;
+            strncpy(tok_mode->last_expr_buffer, tok->cur, size);
+            break;
+        case 0:
+            if (!tok_mode->last_expr_buffer || tok_mode->last_expr_end >= 0) {
+                return 1;
+            }
+            tok_mode->last_expr_buffer = PyMem_Realloc(
+                tok_mode->last_expr_buffer,
+                tok_mode->last_expr_size + size
+            );
+            if (tok_mode->last_expr_buffer == NULL) {
+                tok->done = E_NOMEM;
+                return 0;
+            }
```

Comment:
Isn't this leaking in case `PyMem_Realloc` fails? Ideally we should have something like the below, right?

```c
char *new_last_expr_buffer = PyMem_Realloc(
    tok_mode->last_expr_buffer,
    tok_mode->last_expr_size + size
);
if (new_last_expr == NULL) {
    PyMem_Free(tok_mode->last_expr_buffer);
    tok->done = E_NOMEM;
    return 0;
}
tok_mode->last_expr_buffer = new_last_expr_buffer;